test/serialization/state_saving.cpp

```diff
@@ -28,28 +28,51 @@ using namespace ls;
 using namespace rr;
 using std::filesystem::path;
 
+/**
+ * Suggestions for improving this test suite:
+ *  - Descriptive names for tests.
+ *  - Test one thing at a time.
+ *  - Unit tests work better when you can look at any particular test
+ *    and figure out what its testing which I find hard with
+ *    the "lambda injection" approach.
+ *    - Also, when a test breaks you need to spend a lot of time with the
+ *      debugger to figure out what went wrong.
+ *
```

Comment:
These are reasonable suggestions, but I do think there's a place for high-level tests to supplement the unit tests, because it's too easy to leave something untested when creating unit tests alone, but higher-level end-to-end approaches test everything.  Ideally when something goes wrong, you find it in a unit test, but that doesn't always happen.  I would suggest that as we find bugs in our code that are only found by these higher-level end-to-end tests, that we add unit tests that would have found them.

Reply:
>  there's a place for high-level tests to supplement the unit tests
Absolutely, but the unit tests must actually exist in order to be complemented by integration/end-to-end testing. I think I just wrote in frustration when I couldn't see what was broken let-a-lone what the problem actually was or how to fix it. 

> I would suggest that as we find bugs in our code that are only found by these higher-level end-to-end tests, that we add unit tests that would have found them

Seems like a good suggestion.